File: src/python/20_operacje_na_plikach/zad10.py
"""
Tytul: Znalezienie w folderze wszystkich plikow z danym rozszerzeniem
Tresc: Otrzymujesz dwa napisy. Pierwszy napis to sciezka folderu, a drugi to rozszerzenie szukanych plikow. Znajdz w folderze wszystkie pliki o podanym rozszerzeniu i zapisz ich nazwy w liscie.
Dane wejsciowe: Dwa napisy
Dane wyjsciowe: Lista napisow

"""

import pathlib
import shutil

def skopiuj_pliki_png(sciezka_1, sciezka_2):
    """
    Funkcja kopiuje wszystkie pliki .png z folderu sciezka_1 do folderu sciezka_2.
    """

    folder_1 = pathlib.Path(sciezka_1)
    folder_2 = pathlib.Path(sciezka_2)

    for plik in folder_1.glob("**/*.png"):
        shutil.copy(str(plik), str(folder_2))

def test_skopiuj_pliki_png():

    # stworz foldery testowe
    pathlib.Path("test_1").mkdir(parents=True, exist_ok=True)
    pathlib.Path("test_2").mkdir(parents=True, exist_ok=True)

    # stworz pliki testowe
    pathlib.Path("test_1/test_1.png").touch()
    pathlib.Path("test_1/test_2.png").touch()

    # skopiuj pliki
    skopiuj_pliki_png("test_1", "test_2")

    # sprawdz czy pliki zostaly skopiowane
    assert pathlib.Path("test_2/test_1.png").exists()
    assert pathlib.Path("test_2/test_2.png").exists()
    assert pathlib.Path("test_1/test_1.png").exists()
    assert pathlib.Path("test_1/test_2.png").exists()

    # usun foldery testowe
    shutil.rmtree("test_1")
    shutil.rmtree("test_2")

if __name__ == "__main__":

    test_skopiuj_pliki_png()

File: src/cpp/20_operacje_na_plikach/zad10.cpp
/*
Tytul: Skopiuj pliki.
Tresc zadania: Otrzymujesz dwa napisy reprezentujace sciezki do folderow. Skopiuj wszystkie pliki PNG z pierwszego folderu do drugiego folderu.
Dane wejsciowe: Dwa napisy reprezentujace sciezki do folderow.
Dane wyjsciowe: Brak.

*/
#include <cassert>
#include <experimental/filesystem>
#include <fstream>
#include <iostream>
#include <string>

namespace filesys = std::experimental::filesystem;

std::string nazwaPliku(const std::string &sciezka) {
  return filesys::path(sciezka).filename();
}

void skopiujPlik(const std::string &sciezka,
                 const std::string &sciezkaDocelowa) {
  try {
    filesys::copy_file(sciezka,
                       sciezkaDocelowa + filesys::path::preferred_separator +
                           nazwaPliku(sciezka),
                       filesys::copy_options::overwrite_existing);
  } catch (filesys::filesystem_error &e) {
    std::cout << e.what() << std::endl;
  }
}

std::string znajdzRozszerzenie(std::string sciezka) {

  filesys::path obiekt(sciezka);

  if (obiekt.has_extension())
    return obiekt.extension().string();

  return "";
}

void skopiujPliki(const std::string &sciezka,
                  const std::string &sciezkaDocelowa) {

  for (const auto plik : filesys::directory_iterator(sciezka)) {
    if (znajdzRozszerzenie(plik.path()) == ".png")
      skopiujPlik(plik.path(), sciezkaDocelowa);
  }
}

void testSkopiujPliki() {

  // stworz foldery test1 i folder test2
  std::string sciezka1 = "test1";
  std::string sciezka2 = "test2";
  filesys::create_directory(sciezka1);
  filesys::create_directory(sciezka2);

  // stworz pliki test1/plik1.png i test1/plik2.png
  std::string sciezkaPliku1 =
      sciezka1 + filesys::path::preferred_separator + "plik1.png";
  std::string sciezkaPliku2 =
      sciezka1 + filesys::path::preferred_separator + "plik2.png";
  std::ofstream plik1(sciezkaPliku1);
  std::ofstream plik2(sciezkaPliku2);
  plik1.close();
  plik2.close();

  // skopiuj pliki z test1 do test2
  skopiujPliki(sciezka1, sciezka2);

  // sprawdz czy pliki zostaly skopiowane
  assert(filesys::exists(sciezka2 + filesys::path::preferred_separator +
                         "plik1.png"));
  assert(filesys::exists(sciezka2 + filesys::path::preferred_separator +
                         "plik2.png"));

  // usun foldery test1 i test2
  filesys::remove_all(sciezka1);
  filesys::remove_all(sciezka2);
}

int main() {

  testSkopiujPliki();

  return 0;
}

